pkg/ssh/connection_golang.go

```diff
@@ -289,11 +290,24 @@ func ValidateAndConfigure(uri *url.URL, iden string, insecureIsMachineConnection
 			keyFilePath := filepath.Join(homedir.Get(), ".ssh", "known_hosts")
 			known, err := knownhosts.New(keyFilePath)
 			if err != nil {
-				if errors.Is(err, os.ErrNotExist) {
-					logrus.Warn("please create a known_hosts file. The next time this host is connected to, podman will add it to known_hosts")
-					return nil
+				if !errors.Is(err, os.ErrNotExist) {
+					return err
+				}
+				keyDir := path.Dir(keyFilePath)
+				if _, err := os.Stat(keyDir); errors.Is(err, os.ErrNotExist) {
+					if err := os.Mkdir(keyDir, 0o700); err != nil {
```

Comment:
FYI, this broke us in OCP where we run `podman --remote` commands in an unprivileged pod running as a made-up uid without a proper homedir:

https://github.com/coreos/coreos-ci-lib/pull/144

We hacked around this by just defining a bogus `HOME`. I wonder if a better fix here would've been to treat `EPERM` from this `os.Mkdir` as non-fatal and just keep warning like it did before.

Reply:
Miloslav actually [pointed this out above](https://github.com/containers/common/pull/1216#issuecomment-1300752631):

> If ~/.ssh does not exist, there is now an error that includes the full path. Good, the user asked for a full path and got a full path — but now that’s not a warning, but a hard error.

Reply:
(A historical note: My concern was about `.ssh` not existing, and the PR _was_ updated to handle it the way I expected it to.)

---

> I wonder if a better fix here would've been to treat `EPERM` from this `os.Mkdir` as non-fatal and just keep warning like it did before.

I’m really rather uncomfortable with that. _All_ of the security of ssh relies on being able to trust the server’s public key. The trust-on-first-use style of operation is _already_ a pretty big compromise of security for ease of use (and a much bigger decrease of security than originally intended nowadays, with ephemeral containers where a much larger proportion of uses are “first use”).

So failing on a complete inability to use `known_hosts` seems generally desirable to me.

Ideally, I think it should be possible for users to specify the expected public key directly in `podman system connection add`; then everything could work without having to touch the global state in `~/.ssh`.

Reply:
> (and a much bigger decrease of security than originally intended nowadays, with ephemeral containers where a much larger proportion of uses are “first use”).

Yes indeed, this is actually our primary use case for podman remote. We're constantly (re-)trusting on first use. Though some of our uses of podman remote are over a more trusted infrastructure, some of it isn't.

> So, failing on a complete inability to use `known_hosts` seems generally desirable to me.
>
> Ideally, I think it should be possible for users to specify the expected public key directly in `podman system connection add`; then everything could work without having to touch the global state in `~/.ssh`.

I commented here initially to signal a break in case more people show up (podman v4.4 just entered f37 a few days ago). In practice, I think it's not common to do `podman remote` things from an OpenShift pod like we do (since the more natural thing to keep your container workloads within k8s).

But yeah, we should probably tighten this up anyway. It wouldn't be too hard for our wrapper code to first add the known hosts entries (from e.g. a configmap) before calling podman remote. The harder part is reprovisioning these remote machines with the same host keys each time ([secret provisioning](https://coreos.github.io/ignition/operator-notes/#secrets) in general is an issue). It'd probably be easier I think to just have them naturally rotate everytime we reprovision and update the configmap or equivalent. So we'd at least reduce it to a single TOFU per reprovision.

Reply:
> I commented here initially to signal a break in case more people show up (podman v4.4 just entered f37 a few days ago).

Thanks! There has been at least one other report about users who are now facing failures.

To an extent the tightened security is intentional and desirable, but both reports suggest the possible need to add more features to allow easier management of the host keys or the like. So it’s definitely useful to keep the conversation going.

Just, please, file RFEs you eventually decide would help, if any, as individual issues, probably on the Podman repo.